dbms/src/Flash/Planner/Plans/PhysicalExpand.cpp

```diff
@@ -0,0 +1,132 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/FailPoint.h>
+#include <Common/Logger.h>
+#include <Common/TiFlashException.h>
+#include <DataStreams/ExpressionBlockInputStream.h>
+#include <DataTypes/DataTypeNullable.h>
+#include <Flash/Coprocessor/DAGExpressionAnalyzer.h>
+#include <Flash/Coprocessor/DAGPipeline.h>
+#include <Flash/Pipeline/Exec/PipelineExecBuilder.h>
+#include <Flash/Planner/FinalizeHelper.h>
+#include <Flash/Planner/PhysicalPlanHelper.h>
+#include <Flash/Planner/Plans/PhysicalExpand.h>
+#include <Interpreters/Context.h>
+#include <Operators/ExpressionTransformOp.h>
+#include <Operators/FilterTransformOp.h>
+#include <fmt/format.h>
+
+namespace DB
+{
+PhysicalPlanNodePtr PhysicalExpand::build(
+    const Context & context,
+    const String & executor_id,
+    const LoggerPtr & log,
+    const tipb::Expand & expand,
+    const PhysicalPlanNodePtr & child)
+{
+    assert(child);
+
+    child->finalize();
+
+    if (unlikely(expand.grouping_sets().empty()))
+    {
+        //should not reach here
+        throw TiFlashException("Expand executor without grouping sets", Errors::Planner::BadRequest);
+    }
+
+    DAGExpressionAnalyzer analyzer{child->getSchema(), context};
+    ExpressionActionsPtr before_expand_actions = PhysicalPlanHelper::newActions(child->getSampleBlock());
+    ExpressionActionsPtr expand_actions_itself = PhysicalPlanHelper::newActions(child->getSampleBlock());
+
+    auto grouping_sets = analyzer.buildExpandGroupingColumns(expand, before_expand_actions);
+    auto expand_action = ExpressionAction::expandSource(grouping_sets);
+    expand_actions_itself->add(expand_action);
+
+    // construct sample block.
+    NamesAndTypes expand_output_columns;
+    auto child_header = child->getSchema();
+    for (const auto & one : child_header)
+    {
+        expand_output_columns.emplace_back(one.name, expand_action.expand->isInGroupSetColumn(one.name) ? makeNullable(one.type) : one.type);
+    }
+    expand_output_columns.emplace_back(expand_action.expand->grouping_identifier_column_name, expand_action.expand->grouping_identifier_column_type);
+
+    auto physical_expand = std::make_shared<PhysicalExpand>(
+        executor_id,
+        expand_output_columns,
+        log->identifier(),
+        child,
+        expand_action.expand,
+        expand_actions_itself,
+        Block(expand_output_columns));
+
+    return physical_expand;
+}
+
+
+void PhysicalExpand::expandTransform(DAGPipeline & child_pipeline)
+{
+    String expand_extra_info = fmt::format("expand, expand_executor_id = {}", execId());
+    FmtBuffer fb;
+    fb.append(": grouping set ");
+    shared_expand->getGroupingSetsDes(fb);
+    expand_extra_info.append(fb.toString());
```

Comment:
I prefer to make `getGroupingSetsDes` return string directly.
```suggestion
    String expand_extra_info = fmt::format("expand, expand_executor_id = {}: grouping set {}", execId(), shared_expand->getGroupingSetsDes());
```